scipy/stats/_multicomp.py

```diff
@@ -0,0 +1,253 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from scipy import stats
+from scipy.stats._qmc import check_random_state
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from scipy._lib._util import SeedType
+    from typing import Literal
+
+
+__all__ = [
+    'dunnett'
+]
+
+
+@dataclass
+class DunnettResult:
+    statistic: np.ndarray
+    pvalue: np.ndarray
+
+
+def dunnett(
+    *samples: npt.ArrayLike, control: npt.ArrayLike,
+    alternative: Literal['two-sided', 'less', 'greater'] = "two-sided",
+    random_state: SeedType = None
+) -> DunnettResult:
+    """Dunnett's test.
+
+    Parameters
+    ----------
+    sample1, sample2, ... : 1D array_like
+        The sample measurements for each experiment group.
+    control : 1D array_like
+        The sample measurements for the control group.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available (default is 'two-sided'):
+
+        * 'two-sided': the means of the distributions underlying the samples
+          and control are unequal.
+        * 'less': the means of the distribution underlying the samples
+          is less than the mean of the distribution underlying the control.
+        * 'greater': the means of the distribution underlying the
+          samples is greater than the mean of the distribution underlying
+          the control.
+    random_state : {None, int, `numpy.random.Generator`}, optional
+        If `random_state` is an int or None, a new `numpy.random.Generator` is
+        created using ``np.random.default_rng(random_state)``.
+        If `random_state` is already a ``Generator`` instance, then the
+        provided instance is used.
+
+    Returns
+    -------
+    res : DunnettResult
+        An object containing attributes:
+
+        statistic : scalar or ndarray
+            The t-statistic for the test.  For 1-D inputs a scalar is
+            returned.
+        pvalue : scalar ndarray
+            The p-value for the test.
```

Comment:
Dunnett's test does not give out a single p-value. Rather, each sample group is compared against the control and gives out a p-value for itself. So for `n` comparisons (for `n = len(samples)`), you'll have `n` t-statistics and `n` p-values. This is similar to `tukey_hsd()` which returns matrices of test statistics and p-values for the many pairwise comparisons that it does.

The non-infinite bound (lower or upper or both, depending on the `alternative`) for the multivariate-t is not the vector of the various t-statistics lumped together. Rather `n` separate integrations are done with the `np.full(t_stat_for_this_group, n_groups)` being the non-infinite bound vector. The idea is that you want to find the probability of the null hypothesis (sample means equaling the control mean) creating at least one sample group at least as extreme as the test statistic.

Reply:
Oops some leftover, I rephrased based on what's in `tukey_hsd`

Reply:
Ah, got it. I thought the code might also be computing a scalar, but I see the broadcasting in the `multivariate_t.cdf()` call now. I confirm that this was just a documentation issue.